File: 3rd Sem/NM/python/derivativeusingbackwarddiff.py
def main():
    # Input number of data points
    n = int(input("Enter number of data points: "))

    # Input x values
    x = list(map(float, input("Enter x values: ").split()))

    # Initialize 2D list for difference table
    y = []
    print("Enter y values:")
    for i in range(n):
        row = list(map(float, input(f"For x[{i}]: ").split()))
        # Fill the rest of the row with zeros to hold future differences
        row += [0] * (n - len(row))
        y.append(row)

    # Step size (assumes equally spaced)
    h = x[1] - x[0]

    # Create backward difference table
    for j in range(1, n):
        for i in range(n - 1, j - 1, -1):
            y[i][j] = y[i][j - 1] - y[i - 1][j - 1]

    # Calculate first and second derivative at last point
    b1 = y[n - 1][1] / h
    b2 = (y[n - 1][1] - y[n - 1][2] / 2) / h

    # Output
    print(f"First Derivative: {b1:.6f}")
    print(f"Second Derivative: {b2:.6f}")

# Run the main function
main()

File: 3rd Sem/NM/C-program/derivativeusingbackwarddiff.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    int n;
    printf("Enter number of data points: ");
    scanf("%d", &n);

    double *x = (double *)malloc(n * sizeof(double));
    double **y = (double **)malloc(n * sizeof(double *));
    
    printf("Enter x values: ");
    for (int i = 0; i < n; i++) {
        scanf("%lf", &x[i]);
    }

    printf("Enter y values:\n");
    for (int i = 0; i < n; i++) {
        y[i] = (double *)malloc(n * sizeof(double));
        printf("For x[%d]: ", i);
        for (int j = 0; j < n; j++) {
            scanf("%lf", &y[i][j]);
        }
    }

    double h = x[1] - x[0];

    for (int j = 1; j < n; j++) {
        for (int i = n - 1; i >= j; i--) {
            y[i][j] = y[i][j - 1] - y[i - 1][j - 1];
        }
    }

    double b1 = y[n - 1][1] / h;
    double b2 = (y[n - 1][1] - (y[n - 1][2] / 2)) / h;

    printf("First Derivative: %.6f\n", b1);
    printf("Second Derivative: %.6f\n", b2);

    for (int i = 0; i < n; i++) {
        free(y[i]);
    }
    free(y);
    free(x);

    return 0;
}
